### PR TITLE
Reset provisioning timeout when we hear from mesos

### PR DESCRIPTION
**Pull request by alenkacz**
JIRA issues: MARATHON-8606

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala [hunk 1]
```diff
@@ -278,7 +278,9 @@ private class TaskLauncherActor(
 
         // Only instances that scheduled or provisioned have not seen a Mesos update. The provision timeouts waits for
         // any Mesos update. Thus we can safely kill the provision timeout in all other cases, even on a TASK_FAILED.
-        if (!instance.isProvisioned && !instance.isScheduled) {
+        // with stable ids, TASK_FAILED ends up yielding instance.isScheduled typically in case of goal: Running
+        // because of that we have to handle instance becoming terminal explicitly
+        if ((!instance.isProvisioned && !instance.isScheduled) || instance.state.condition.isTerminal) {
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala]:
```suggestion
        def isLaunchedOrRescheduledOnFailure(instance: Instance): Boolean = (!instance.isProvisioned && !instance.isScheduled) || instance.state.condition.isTerminal
        if (isLaunchedOrRescheduledOnFailure(instance)) {
```

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala [hunk 2]
```diff
@@ -278,7 +278,9 @@ private class TaskLauncherActor(
 
         // Only instances that scheduled or provisioned have not seen a Mesos update. The provision timeouts waits for
         // any Mesos update. Thus we can safely kill the provision timeout in all other cases, even on a TASK_FAILED.
-        if (!instance.isProvisioned && !instance.isScheduled) {
+        // with stable ids, TASK_FAILED ends up yielding instance.isScheduled typically in case of goal: Running
+        // because of that we have to handle instance becoming terminal explicitly
+        if (!instance.isProvisioned && provisionTimeouts.contains(instanceId)) {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala]:
```suggestion
        if (!instance.isProvisioned) {
```
I think `provisionTimeouts.get(instanceId).foreach(_.cancel())` already does the check 🙂

File: src/test/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActorTest.scala [hunk 3]
```diff
@@ -406,5 +406,30 @@ class TaskLauncherActorTest extends AkkaUnitTest with Eventually {
...
+        launcherRef.underlyingActor.provisionTimeouts.contains(f.scheduledInstance.instanceId) shouldBe (true)
+      }
+
+      When("the task fails after being provisioned")
+      val op = mock[InstanceOp]
+      op.instanceId returns f.provisionedInstance.instanceId
+      launcherRef ! TaskStatusUpdateTestHelper.failed(f.provisionedInstance).wrapped
+
+      Then("the provisioning timeout is removed")
+      launcherRef.underlyingActor.provisionTimeouts.contains(f.scheduledInstance.instanceId) shouldBe (false)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActorTest.scala]:
```suggestion
      launcherRef.underlyingActor.provisionTimeouts.keys should contain(f.scheduledInstance.instanceId)
```

File: src/test/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActorTest.scala [hunk 4]
```diff
@@ -406,5 +406,30 @@ class TaskLauncherActorTest extends AkkaUnitTest with Eventually {
...
+      val launcherRef = createLauncherRef()
+      launcherRef ! RateLimiter.DelayUpdate(f.app.configRef, None)
+      Mockito.when(instanceOpFactory.matchOfferRequest(m.any())).thenReturn(f.launchResult)
+
+      val promise = Promise[MatchedInstanceOps]
+      val offer = MarathonTestHelper.makeBasicOffer().build()
+      launcherRef ! ActorOfferMatcher.MatchOffer(offer, promise)
+
+      eventually {
+        launcherRef.underlyingActor.provisionTimeouts.contains(f.scheduledInstance.instanceId) shouldBe (true)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActorTest.scala]:
```suggestion
        launcherRef.underlyingActor.provisionTimeouts.keys should contains(f.scheduledInstance.instanceId)
```

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala [hunk 5]
```diff
@@ -278,7 +278,9 @@ private class TaskLauncherActor(
 
         // Only instances that scheduled or provisioned have not seen a Mesos update. The provision timeouts waits for
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala]: I believe this comment is not correct anymore: 
```suggestion
        // Only instances that are provisioned have not seen a Mesos update yet. The provision timeout waits for
```

File: src/test/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActorTest.scala [hunk 6]
```diff
@@ -406,5 +406,30 @@ class TaskLauncherActorTest extends AkkaUnitTest with Eventually {
       Then("the instance is not rescheduled")
       verify(instanceTracker, never).forceExpunge(any[Instance.Id])
     }
+
+    "reset provisioning timeout after instance failed" in new Fixture {
+      Given("a provisioned instance")
+      Mockito.when(instanceTracker.instancesBySpecSync).thenReturn(InstanceTracker.InstancesBySpec.forInstances(f.scheduledInstance))
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActorTest.scala]: Don't resist it ;)
```suggestion
      instanceTracker.instancesBySpecSync returns InstanceTracker.InstancesBySpec.forInstances(f.scheduledInstance)
```

File: src/test/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActorTest.scala [hunk 7]
```diff
@@ -406,5 +406,30 @@ class TaskLauncherActorTest extends AkkaUnitTest with Eventually {
...
+
+      val launcherRef = createLauncherRef()
+      launcherRef ! RateLimiter.DelayUpdate(f.app.configRef, None)
+      Mockito.when(instanceOpFactory.matchOfferRequest(m.any())).thenReturn(f.launchResult)
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActorTest.scala]:
```suggestion
      instanceOpFactory.matchOfferRequest(m.any()) returns f.launchResult
```

**Pull request merged**